lib/handleRequest.js

```diff
@@ -125,30 +129,72 @@ function validationCompleted (request, reply, validationErr) {
 function preHandlerCallback (err, request, reply) {
   if (reply.sent) return
 
-  if (err != null) {
-    reply[kReplyIsError] = true
-    reply.send(err)
-    return
+  const context = request[kRouteContext]
+
+  if (!tcHasSubscribers() || context[kFourOhFourContext] === null) {
+    preHandlerCallbackInner(err, request, reply)
+  } else {
+    const store = {
+      request,
+      reply,
+      async: false,
+      route: {
+        url: context.config.url,
+        method: context.config.method
+      }
+    }
+    channels.start.runStores(store, preHandlerCallbackInner, undefined, err, request, reply, store)
   }
+}
 
-  let result
+function preHandlerCallbackInner (err, request, reply, store) {
+  const context = request[kRouteContext]
 
   try {
-    result = request[kRouteContext].handler(request, reply)
-  } catch (err) {
-    reply[kReplyIsError] = true
-    reply.send(err)
-    return
-  }
+    if (err != null) {
+      reply[kReplyIsError] = true
+      reply.send(err)
+      if (store) {
+        store.error = err
+        channels.error.publish(store)
+      }
+      return
+    }
 
-  if (result !== undefined) {
-    if (result !== null && typeof result.then === 'function') {
-      wrapThenable(result, reply)
-    } else {
-      reply.send(result)
+    let result
+
+    try {
+      result = context.handler(request, reply)
+    } catch (err) {
+      if (store) {
+        store.error = err
+        channels.error.publish(store)
+      }
+
+      reply[kReplyIsError] = true
+      reply.send(err)
+      return
+    }
+
+    if (result !== undefined) {
+      if (result !== null && typeof result.then === 'function') {
+        wrapThenable(result, reply, store)
+      } else {
+        reply.send(result)
+      }
     }
+  } finally {
+    if (store) channels.end.publish(store)
   }
 }
 
+function tcHasSubscribers () {
+  return channels.start.hasSubscribers ||
+    channels.end.hasSubscribers ||
+    channels.asyncStart.hasSubscribers ||
+    channels.asyncEnd.hasSubscribers ||
+    channels.error.hasSubscribers
```

Comment:
@Qard do you agree that there should be a `tracingChannel.hasSubscribers() === true` method in core so that implementations don't have to maintain a function like this one?

Reply:
Yes, just landed it the other day with the change to enable early exits. It's not in dc-polyfill yet though.